### PR TITLE
Build the toolbox and client containers

### LINKED ISSUE #411: automate the manual builds (rook-client and toolbox)
Both the `rook-client` and `toolbox` containers are currently build and published manually (I believe it's just those two).  We could improve this with an automated solution that:
- only runs as part of our build scripts when needed (such as at real release time), so these containers aren't built every time a dev builds
- do not rely on a hard coded `rook` version in their `Dockerfile` to download the `rook` binary from github.  Instead they should use the locally built `rook` binary, which would allow these containers to be versioned along with the other release artifacts (`rook`, `rookd`, `rook-operator`)

### PR DESCRIPTION
**Pull request by travisn**
Fixes #411

### CODE REVIEW

File: build/release/containers.sh [hunk 1]
```diff
@@ -52,16 +52,60 @@ build() {
...
 ENTRYPOINT ["/usr/bin/rookd"]
 EOF
-    build_artifact $os $arch $tmpdir rook/rookd rookd
-    rm -fr $tmpdir
+    build_artifact $os $arch $tmpdir1 rook/rookd rookd
+    rm -fr $tmpdir1
+
+    echo "Building the rook-client container"
+    tmpdir2=$(mktemp -d)
+    trap "rm -fr $tmpdir2" EXIT
```
Comment (bassam) [inline build/release/containers.sh]: you dont need multiple traps. just one.

File: build/release/containers.sh [hunk 2]
```diff
@@ -52,16 +52,60 @@ build() {
...
+    cat <<EOF > $tmpdir2/Dockerfile
+FROM ubuntu
```
Comment (bassam) [inline build/release/containers.sh]: does this need to be ubuntu? can we just use a scratch container or alpine? Also with #614 you're going to run into conflicts and need to use a cross platform base image

File: build/release/containers.sh [hunk 3]
```diff
@@ -52,16 +52,60 @@ build() {
...
+    build_artifact $os $arch $tmpdir2 rook/rook-client rook
+    rm -fr $tmpdir2
+
+    echo "Building the toolbox container"
+    tmpdir3=$(mktemp -d)
+    cp toolbox/entrypoint.sh $tmpdir3
+    trap "rm -fr $tmpdir3" EXIT
+    cat <<EOF > $tmpdir3/Dockerfile
+FROM ubuntu:xenial
+RUN apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 08b73419ac32b4e966c1a330e84ac2c0460f3994 \
```
Comment (bassam) [inline build/release/containers.sh]: I wish we used the same version of ceph that rook is using.

File: build/release/containers.sh (hunk 1 above)
Comment (travisn) [inline build/release/containers.sh]: sure i'll combine to a single trap

File: build/release/containers.sh (hunk 3 above)
Comment (travisn) [inline build/release/containers.sh]: right, i'll update it to pull luminous if the build flag is set as such.

File: build/release/containers.sh (hunk 2 above)
Comment (travisn) [inline build/release/containers.sh]: the walkthrough right now assumes extra tools are in the container with the client, but that's not right. The walkthrough should really use the toolbox for anything besides the client. I'll look at updating that so we can simplify the client container

**Comment by bassam** [comment]
also you need to ensure that new docker images are cleaned up at the end of every build

**Comment by travisn** [comment]
ok will add to the `cleanup` function

File: build/release/containers.sh [hunk 4]
```diff
@@ -52,16 +52,60 @@ build() {
...
+    && echo "deb http://download.ceph.com/debian-kraken/ xenial main" > /etc/apt/sources.list.d/ceph-kraken.list \
+    && apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -yq --no-install-recommends \
```
Comment (jbw976) [inline build/release/containers.sh]: we will get the benefits of image layer caching during this build right?  I'm hoping these downloads don't get run every single build.

Comment (travisn) [inline build/release/containers.sh]: we will definitely get the benefit locally during development. On jenkins i'm not sure if the cleanup will remove all layers

File: build/release/containers.sh (hunk 2 above)
Comment (travisn) [inline build/release/containers.sh]: the change to base the client image on alpine is turning into a big change to remove a dependency on bash. I think it best to split out that change into a separate PR and keep the container on ubuntu for this PR.

**Comment by bassam** [comment]
do we still need the toolbox directory?

**Comment by travisn** [comment]
i was tempted to merge it into demo/kubernetes. The entrypoint.sh just needs a place to live. And it does seem a bit odd to have all our rook yamls in the demo folder. @jbw976 ?

**Comment by jbw976** [comment]
I can't think of a better place for the toolbox's entrypoint script to live than in the toolbox folder.  While we're on the topic, I've always wondered why we inline our Dockerfiles directly into the `build/release/containers.sh` script, instead of having them separate, for example /toolbox/Dockerfile.  I've seen a couple questions regarding where our rookd container Dockerfile is and I haven't been quite sure why it's inlined into the build scripts.

**Comment by bassam** [comment]
we need a better place for our yaml files. we can certainly use external Dockerfiles if needed. in the case of rookd we still need to munge them since the base image can change.

**Comment by travisn** [comment]
ok, we'll move the yaml files around in a future PR
